crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,532 @@
+        // have length of at least two. If there are fewer than two items in `__all__`,
+        // it is impossible for them *not* to compare equal here:
+        if sorted_items == self.items {
+            return SortedDunderAll::AlreadySorted;
+        }
+        assert!(self.items.len() >= 2);
+
+        // As well as the "items" in the `__all__` definition,
+        // there is also a "prelude" and a "postlude":
+        //  - Prelude == the region of source code from the opening parenthesis
+        //    (if there was one), up to the start of the first element in `__all__`.
+        //  - Postlude == the region of source code from the end of the last
+        //    element in `__all__` up to and including the closing parenthesis
+        //    (if there was one).
+        let prelude_end = {
+            // We should already have returned by now if there are 0 items:
+            // see earlier comments in this function
+            let first_item = self
+                .items
+                .first()
+                .expect("Expected there to be at least two items in the list");
+            let first_item_line_offset = locator.line_start(first_item.start());
+            if first_item_line_offset == locator.line_start(self.start()) {
+                first_item.start()
+            } else {
+                first_item_line_offset
+            }
+        };
+        let (needs_trailing_comma, postlude_start) = {
+            // We should already have returned by now if there are 0 items:
+            // see earlier comments in this function
+            let last_item = self
+                .items
+                .last()
+                .expect("Expected there to be at least two items in the list");
+            let last_item_line_offset = locator.line_end(last_item.end());
+            if last_item_line_offset == locator.line_end(self.end()) {
+                (false, last_item.end())
+            } else {
+                (true, last_item_line_offset)
+            }
+        };
+        let mut prelude = locator
+            .slice(TextRange::new(self.start(), prelude_end))
+            .to_string();
+        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));
+
+        let joined_items = if self.multiline {
+            let indentation = stylist.indentation();
+            let newline = stylist.line_ending().as_str();
+            prelude = format!("{}{}", prelude.trim_end(), newline);
+            join_multiline_dunder_all_items(
+                &sorted_items,
+                locator,
+                parent,
+                indentation,
+                newline,
+                needs_trailing_comma,
+            )
+        } else {
+            Some(join_singleline_dunder_all_items(&sorted_items, locator))
+        };
+
+        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
+        SortedDunderAll::Sorted(new_dunder_all)
+    }
```

Comment:
We can avoid allocating here because you're constructing a new string below anyway. Sorry, I'm very picky when it comes to allocations because they tend to be expensive (see https://vimeo.com/649009599)


```suggestion
        let mut prelude = Cow::Borrowed(locator.slice(TextRange::new(self.start(), prelude_end)));

        let joined_items = if self.multiline {
            let indentation = stylist.indentation();
            let newline = stylist.line_ending().as_str();
            prelude = Cow::Owned(format!("{}{}", prelude.trim_end(), newline));
            join_multiline_dunder_all_items(
                &sorted_items,
                locator,
                parent,
                indentation,
                newline,
                needs_trailing_comma,
            )
        } else {
            Some(join_singleline_dunder_all_items(&sorted_items, locator))
        };

        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));

        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
        SortedDunderAll::Sorted(new_dunder_all)
    }
```

Reply:
Should `postlude` also be a `Cow`, or just `prelude`?

Reply:
Having done a bit more reading, I can answer my own question: no, `postlude` doesn't need to be a `Cow`, since the type of `postlude` currently is already a borrowed type (`&str`), which is what we want, but the type of `prelude` currently is an owned type (`String`), which is what we don't want, and what using `Cow` fixes for us.

Thanks!

Reply:
bec4a1b714f53fce67b98acaae7b922934537cc2